RULES/cc/cc.go

```diff
@@ -99,26 +97,56 @@ type Library struct {
 	Shared        bool
 	AlwaysLink    bool
 	Toolchain     Toolchain
+
+	multipleToolchains bool
+	toolchainMap       map[string]Library
+	baseOut            core.OutPath
+}
+
+func (lib Library) MultipleToolchains() Library {
+	lib.multipleToolchains = true
+	lib.toolchainMap = make(map[string]Library)
+	lib.baseOut = lib.Out
+	return lib
 }
 
 // Build a Library.
 func (lib Library) Build(ctx core.Context) {
 	toolchain := lib.Toolchain
 	if toolchain == nil {
-		toolchain = &DefaultToolchain
+		toolchain = defaultToolchain()
+	}
+
+	if lib.multipleToolchains {
```

Comment:
I am not sure I get how this is supposed to work. Am I correct that you have a regular lib to start with, then call `.MultipleToolchains` on it, which gives you a new lib with multi-toolchain support. You then call `.WithToolchain` on that new lib with the toolchain you want. Correct?
1) why can't we make all libs have multitoolchain support by default? I.e., set `multipleToolchains` to `true` always
2) Why wouldn't the default lib be placed in the map?

Reply:
yes, correct. But as long as you only use cc.Binary and cc.Library, the right one is selected automatically. In fact WithToolchain could have been private (I never call it in libsupcxx), I made it public in case someone wanted to access a specific Library. I think the libsupcxx targets are a good example of this in action
1) I would have done it that way, but the map must be initialised (`make`) somewhere. This is basically just a constructor. And since you might not want to support multiple toolchains, I made it so that if you don't initialise it, it defaults to only compiling with one. I did not want everyone using Lib to use the constructor, when they might not need to support multiple toolchains.
2) (sorry for updating this while you were reviewing) The default lib is placed in the map and built when we call WithToolchain() on it. The `defaultLib` variable here just copies it from, for example `libsupcxx/io/x86_64/libio.a` to `libsupcxx/io/libio.a`. This is only really useful when the end user wants to specifically `dbt build` this library. If the library is used as a dependency, the `libsupcxx/io/x86_64/libio.a` output is used, and there is nothing special about it

Reply:
I think this is ok. Just one idea: If we initialized the `map` lazily when it is needed for the first time (it is not used from outside), could we get rid of the constructor? Would that be a better solution? I'll leave that to your judgement, since you spent way more time thinking about this.

Reply:
I tried to do that, the problem is that the only other place to initialise it is in Build(), which gets the lib by value. So there I can change the content of the map, but not the map itself. We could use pointers for libs to work around that, but I feel like that'd be a lot more strange from the user's perspective. And we'd still need to have a way to enforce a single toolchain (probably a field in that case). I don't think the added complication is worth it just to change the default option.

go's lack of (enforced) constructors means that the "default option" is forced on you pretty often.

Reply:
Then it's fine.